### ISSUE TITLE
Use GitHub Checks API from our functional test workflow to gate merges

### ISSUE THREAD

**Issue by rynowak** [radius-project/radius#6548]
### Area for Improvement (required)

Functional testing action should block merge until functional tests pass on the **exact** commit being merged. The current state of the art is that we rely on the code-reviewer to enforce this.

### Proposed solution

Hey Folks,

I found a good example blog of someone doing basically exactly what we’re trying to with /ok-to-test. This solution is integration with GitHub Checks. For our solution, it means a PR cannot be merged until the functional tests pass on the exact commit (eg: exactly what we want).

You can find the Blog writeup here: https://josh-ops.com/posts/using-github-checks-api/

You should also look at this post because creating a GitHub App is required: https://josh-ops.com/posts/github-apps/ It probably sounds scary to create a GitHub App, but it’s a similar amount of work to creating a PAT. Now that I know more about it, I feel like we should be using GitHub Apps for more things. The reason why we need to create an app is that the Checks API only works correctly from an App. I tried without creating an App first and it's buggy, GH support says not to do that.

I created a working example of this that doesn’t run any tests, so I could prove out the idea. You can find it here: https://github.com/rynowak/actions-testing There’s no permissions limiting /ok-to-test on my repo, so feel free to try it out.

The basic flow is like this:

- /ok-to-test does a repository dispatch to start the functional test workflow (we already have this)
- Inside the functional test workflow we login as the app ([here](https://github.com/rynowak/actions-testing/blob/main/.github/workflows/functional-tests.yaml#L12))
- Create a Check for the functional tests using the app's token ([here](https://github.com/rynowak/actions-testing/blob/main/.github/workflows/functional-tests.yaml#L48))
- We run the functional tests (we already have this)
- When the functional tests finish we update the check with the results ([here](https://github.com/rynowak/actions-testing/blob/main/.github/workflows/functional-tests.yaml#L66))
- One this runs ones on the repo we can select the check in the branch protection menu and mark it as required.

My example of this **does not** do a good job reporting the results of the functional test run. This is something that needs to be addressed for us to use this.

[AB#9889](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/9889)

**Comment by youngbupark** [comment]
+1 Using GitHub App looks good idea in terms of security and maintenance and eventually, we can remove rad-bot account. 

When GitHub was first developing GitHub App and Check APIs, I recall that App Center was one of the first customers and I was one to build Visual Studio App Center GitHub Apps - https://github.com/marketplace/app-center From what I learned before, GitHub App (OAuth app) is designed for the integration with 3rd party services via webhook. It is beneficial when we need to apply the similar functionality across the multiple repos in multiple orgs like Azure Pipeline. But, looks like GitHub team allows GitHub Action to get the token for GitHub App, so we don't need to have either `rad-bot` account or separate webhook server in github action to use it like a managed identity. (BTW, as far as I know, PAT can call Check API :) )

**Comment by youngbupark** [comment]
Is it possible to get the GitHub app token via action task in the scenario where user creates the PR from forked repo?

**Comment by rynowak** [comment]
Yes. I built an example of this. 

I did more research as well. There are basically two major approaches here:

- Using GitHub environments. GitHub environments are oriented towards application deployments, but they provide a good UI and integrated experience. This works a lot like `/ok-to-test` but it's built into GitHub. This is less work to set up.
- Using a GH app + the checks API. 

Here's a repo I set up: https://github.com/rynowak/actions-testing

Example using environment: https://github.com/rynowak/actions-testing/blob/main/.github/workflows/functional-tests-environment.yaml#L12

Example using `/ok-to-test` with a GH APP: https://github.com/rynowak/actions-testing/blob/main/.github/workflows/functional-tests.yaml
